app/models/run.rb

```diff
@@ -220,28 +216,42 @@ def possible_timesave(timing)
-      PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.realtime_duration_ms) AS percentile99
-      '.squish
-    when Run::GAME
-      'segment_id,
-      STDDEV_POP(segment_histories.gametime_duration_ms) AS standard_deviation,
-      AVG(segment_histories.gametime_duration_ms) AS mean,
-      PERCENTILE_DISC(.5) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS median,
-      PERCENTILE_CONT(.1) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS percentile10,
-      PERCENTILE_CONT(.9) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS percentile90,
-      PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.gametime_duration_ms) AS percentile99
-      '.squish
-    else
-      raise 'Unsupported timing'
-    end
+  def segment_history_stats_query(timing)
+    timing_string = case timing
+                    when Run::REAL
+                      'real'
+                    when Run::GAME
+                      'game'
+                    else
+                      raise 'Unsupported timing'
+                    end
+
+    sql = %Q{
+      SELECT
+        segment_id,
+        STDDEV_POP(segment_histories.#{timing_string}time_duration_ms) AS standard_deviation,
+        AVG(segment_histories.#{timing_string}time_duration_ms) AS mean,
+        PERCENTILE_DISC(.5) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS median,
+        PERCENTILE_CONT(.1) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS percentile10,
+        PERCENTILE_CONT(.9) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS percentile90,
+        PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY segment_histories.#{timing_string}time_duration_ms) AS percentile99 FROM segment_histories
+      INNER JOIN segments ON segments.id = segment_histories.segment_id
+      INNER JOIN runs ON runs.id = segments.run_id
+      LEFT JOIN (
+        SELECT segment_histories.id AS id, segment_histories.attempt_number AS attempt_number, segments.segment_number as segment_number
+        FROM segment_histories
+        INNER JOIN segments ON segments.id = segment_histories.segment_id
+        WHERE segments.run_id = #{id}
+        AND NOT ((segment_histories.#{timing_string}time_duration_ms = 0 OR segment_histories.#{timing_string}time_duration_ms IS NULL))
+      ) AS other_histories ON other_histories.attempt_number = segment_histories.attempt_number AND other_histories.segment_number = segments.segment_number - 1
+      WHERE runs.id = #{id}
+        AND NOT ((segment_histories.#{timing_string}time_duration_ms = 0 OR segment_histories.#{timing_string}time_duration_ms IS NULL))
+        AND (segments.segment_number = 0 OR
+          (other_histories.attempt_number = segment_histories.attempt_number
+          AND other_histories.segment_number = segments.segment_number - 1)
+        )
+      GROUP BY segment_id
+    }.squish
```

Comment:
I'm going to be 100% honest here, this entire query is way beyond me and my poor SQL experience. I'm physically and mentally exhausted right now so maybe my brain not be do working good, but my eyes are just glazed over from trying to build the mental model for what all the joins are doing.

I'll take a look when I'm in a better mental state, but I do think (assuming this is the simplest query for the job) we could use some comments or somesuch to help readers like myself understand what this is doing and why.

Reply:
If it wasn't clear in my PR description, I'm not really jazzed about this, either. It's tough to read and thus tough to maintain. Even with comments, it still won't be great, but I can definitely add them. Additionally, I'm sure not betting the farm on this being 100% correct. I'd like as many eyes on this as possible.